api/state.go

```diff
@@ -4,187 +4,135 @@
+	// modelTag holds the model tag.
+	// It is empty if there is no model tag associated with the connection.
+	modelTag names.ModelTag
+
+	// controllerTag holds the controller's tag once we're connected.
+	controllerTag names.ControllerTag
+
+	// serverVersion holds the version of the API server that we are
+	// connected to.  It is possible that this version is 0 if the
+	// server does not report this during login.
+	serverVersion version.Number
+
+	// hostPorts is the API server addresses returned from Login,
+	// which the client may cache and use for fail-over.
+	hostPorts []network.MachineHostPorts
+
+	// publicDNSName is the public host name returned from Login
+	// which the client can use to make a connection verified
+	// by an officially signed certificate.
+	publicDNSName string
+
+	// facadeVersions holds the versions of all facades as reported by
+	// Login
+	facadeVersions map[string][]int
+
+	// pingFacadeVersion is the version to use for the pinger. This is lazily
+	// set at initialization to avoid a race in our tests. See
+	// http://pad.lv/1614732 for more details regarding the race.
+	pingerFacadeVersion int
+
+	// authTag holds the authenticated entity's tag after login.
+	authTag names.Tag
+
+	// mpdelAccess holds the access level of the user to the connected model.
+	modelAccess string
+
+	// controllerAccess holds the access level of the user to the connected controller.
+	controllerAccess string
+
+	// broken is a channel that gets closed when the connection is
+	// broken.
+	broken chan struct{}
+
+	// closed is a channel that gets closed when State.Close is called.
+	closed chan struct{}
+
+	// loggedIn holds whether the client has successfully logged
+	// in. It's a int32 so that the atomic package can be used to
+	// access it safely.
+	loggedIn int32
+
+	// tag, password, macaroons and nonce hold the cached login
+	// credentials. These are only valid if loggedIn is 1.
+	tag       string
+	password  string
+	macaroons []macaroon.Slice
+	nonce     string
+
+	// serverRootAddress holds the cached API server address and port used
+	// to login.
+	serverRootAddress string
+
+	// serverScheme is the URI scheme of the API Server
+	serverScheme string
+
+	// tlsConfig holds the TLS config appropriate for making SSL
+	// connections to the API endpoints.
+	tlsConfig *tls.Config
+
+	// bakeryClient holds the client that will be used to
+	// authorize macaroon based login requests.
+	bakeryClient *httpbakery.Client
+
+	// proxier is the proxier used for this connection when not nil. If's expected
+	// the proxy has already been started when placing in this var. This struct
+	// will take the responsibility of closing the proxy.
+	proxier jujuproxy.Proxier
+}
+
+// Login implements the Login method of the Connection interface providing authentication
+// using basic auth or macaroons.
+func (st *state) Login(name names.Tag, password, nonce string, ms []macaroon.Slice) error {
```

Comment:
Should we not be removing this function in favour of a login provider?

Reply:
removed, but added a convenience function to export_test.go

Reply:
Actually.. let's leave this in as a convenient method to log in using the user-pass login provider.. there are quite a few tests in apiserver/admin_test.go that test various modes of log in using the username-password/macaroons/nonce.. most of them open a websocket connection then log in after the fact..

Reply:
Longer term we want to clean this up and have the login providers do it. But to make progress, can we add a TODO to refactor.